### PR TITLE
Implement try block expressions

### PR DESCRIPTION
**Pull request by scottmcm**
I noticed that `try` wasn't a keyword yet in Rust 2018, so...

~~Fix​es https://github.com/rust-lang/rust/issues/52604~~ That was fixed by PR https://github.com/rust-lang/rust/pull/53135
cc https://github.com/rust-lang/rust/issues/31436 https://github.com/rust-lang/rust/issues/50412

### CODE REVIEW

**Comment by scottmcm** [comment]
Ok, I think this is in good enough shape to get reviewed.  I've never touched most of these parts of the compiler before, so I suspect there are opportunities for improvement in how I did things :upside_down_face:

File: src/libsyntax_pos/symbol.rs [hunk 1]
```diff
@@ -413,23 +413,30 @@ declare_keywords! {
     (49, Virtual,            "virtual")
     (50, Yield,              "yield")
 
+    // Edition-specific keywords currently in use.
+    (51, Try,                "try") // >= 2018 Edition Only
```
Comment (petrochenkov) [inline src/libsyntax_pos/symbol.rs]: Nit: Keywords are not normally moved into the "used" category until the feature is stable (e.g. `macro`/`yield`/`async` are all "reserved for future use").

File: src/doc/unstable-book/src/language-features/catch-expr.md [hunk 2]
```diff
@@ -6,22 +6,24 @@ The tracking issue for this feature is: [#31436]
 
 ------------------------
 
-The `catch_expr` feature adds support for a `catch` expression. The `catch`
-expression creates a new scope one can use the `?` operator in.
+The `catch_expr` feature adds support for `try` blocks. A `try`
```
Comment (Centril) [inline src/doc/unstable-book/src/language-features/catch-expr.md]: Should probably rename this to `try_expr`?

**Review by nikomatsakis — APPROVED** [review]
Very nice!

**Comment by nikomatsakis** [comment]
One question: do we want to provide a kind of smoother path for people using `do catch` today? e.g., a nice error message?

**Comment by nikomatsakis** [comment]
Or do we just not care :)

**Comment by Centril** [comment]
@nikomatsakis Here's a sourcegraph query for it, https://sourcegraph.com/search?q=repogroup:crates+case:yes++%5Cbdo%5Cs%2Bcatch%5Cb+max:400

Most of these occurrences are irrelevant and so I think we should not care.

**Comment by Centril** [comment]
Also; having the feature called `catch_expr` for a `try { .. }` block will be confusing, so let's change that bit  :)

**Comment by scottmcm** [comment]
@nikomatsakis Well, it was easy, so I just [did it](https://github.com/rust-lang/rust/pull/52602/files#diff-3a305731b50fbd0843e86a9395740eceR16).

@Centril Agreed; switched to `try_blocks`. (The compiler seems to [use `Try` for `?`](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/enum.ExprKind.html#variant.Try) in some places, so I decided for something more explicit than `try_expr`.)

**Comment by kennytm** [comment]
@bors r-

Merge conflict with `src/librustc/lib.rs`.

**Comment by scottmcm** [comment]
[Updated a few tests](https://github.com/rust-lang/rust/pull/52602/commits/d6ae6c4e1ac10584d9e20628dd16c4562972f5df) to work with NLL and the new error messages; please double-check and re-r+.

File: src/test/compile-fail/try-block-bad-lifetime.rs [hunk 3]
```diff
@@ -12,18 +12,22 @@
...
         // result variable
-        let _result: Result<(), &str> = try {
+        let result: Result<(), &str> = try {
             let my_string = String::from("");
             let my_str: & str = & my_string;
             //~^ ERROR `my_string` does not live long enough
             Err(my_str) ?;
             Err("") ?;
         };
+        do_something_with(result);
```
Comment (nikomatsakis) [inline src/test/compile-fail/try-block-bad-lifetime.rs]: seems fine; I usually just use `drop(result)` for this purpose :)

File: src/test/compile-fail/try-block-maybe-bad-lifetime.rs [hunk 4]
```diff
@@ -12,6 +12,9 @@
 
 #![feature(try_blocks)]
 
+#![inline(never)]
+fn do_something_with<T>(_x: T) {}
+
 // This test checks that borrows made and returned inside try blocks are properly constrained
```
Comment (nikomatsakis) [inline src/test/compile-fail/try-block-maybe-bad-lifetime.rs]: really we should make these UI tests, not compile-fail

Comment (nikomatsakis) [reply src/test/compile-fail/try-block-maybe-bad-lifetime.rs]: feel up for that? Also, I'd prefer a directory named something like `src/test/ui/rfc123-try-block`

Comment (nikomatsakis) [reply src/test/compile-fail/try-block-maybe-bad-lifetime.rs]: or just `try-block`

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by Centril** [comment]
It also seems important to this PR to decide what to do with `try!(expr)` in 2018. The question is currently unresolved in #50412. Presumably we should add some tests to ensure the expected behavior once we do decide what we want.

**Comment by scottmcm** [comment]
@Centril There's definitely more work to do -- such as prohibiting a following `catch` -- but I don't think it needs to be this PR specifically.  This one is pretty well contained to just being a rename right now, and I'd like to keep it that way to separate out the work.

Perhaps it would be worth a separate issue for "`try`-related things that need decisions and/or implementing before the edition", so that it can go in one of the milestones?

**Comment by Centril** [comment]
@scottmcm  Ah; that's fair enough.

Note to self: On the issue of a following `catch`, I'd also like to make a following `match`, `else`, and `finally` token forbidden. (we could make any following keyword + `finally` + `catch` verboten to be extra-extra-safe).

I think we can track `try`-related-things on #50412. :)

**Comment by scottmcm** [comment]
Ah, right; I forgot you'd already made a separate issue for it. :+1:

**Comment by nikomatsakis** [comment]
Well @scottmcm you wanted to wait until EP2 or something, right?

**Comment by scottmcm** [comment]
@nikomatsakis Yeah, this isn't edition-critical and I suspect it will break toolstate.  I'll make the suggested changes, but not soon -- I'm getting on a plane in a few hours.

**Comment by scottmcm** [comment]
@nikomatsakis Ok, that was a more complex rebase than expected, but things should be good now.

**Review by nikomatsakis — APPROVED** [review]

**Comment by kennytm** [comment]
@bors retry 

```
An error occurred while generating the build script.
```

**Comment by scottmcm** [comment]
I'm cursed today:

> [01:31:46] fatal: unable to access 'https://github.com/servo/servo/': Could not resolve host: github.com

https://api.travis-ci.org/v3/job/419489535/log.txt

**Comment by alexcrichton** [comment]
As a heads up, we've got a pretty serious issue which may be a bit of a roadblock for this. Currently the compiler doesn't actually warn about usages of the `try` identifier on the 2015 edition, but that's tracked by https://github.com/rust-lang/rust/issues/53077 and fix for it is at https://github.com/rust-lang/rust/pull/53685. Unfortunately though the primary use of the `try` keyword, the `try!` macro in the standard library, suffers from https://github.com/rust-lang/rust/issues/53686.

For the edition we'll need to have at least some warning for users of the `try!` macro to rustfix into `r#try` or suggest `?`, but currently no warnings are emitted which can make for a difficult edition transition experience. Some eyes and/or thoughts on https://github.com/rust-lang/rust/issues/53686 would be greatly appreciated!

**Pull request merged**